sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java

```diff
@@ -313,27 +315,24 @@ public FileBasedWriteOperation(FileBasedSink<T> sink) {
           sink.getBaseOutputFilenameProvider(), new TemporaryDirectoryBuilder()));
     }
 
-    private static class TemporaryDirectoryBuilder
+    @VisibleForTesting
+    static class TemporaryDirectoryBuilder
         implements SerializableFunction<String, String> {
       // The intent of the code is to have a consistent value of tempDirectory across
       // all workers, which wouldn't happen if now() was called inline.
       Instant now = Instant.now();
 
       @Override
       public String apply(String baseOutputFilename) {
-        try {
-          IOChannelFactory factory = IOChannelUtils.getFactory(baseOutputFilename);
-          Path baseOutputPath = factory.toPath(baseOutputFilename);
-          return baseOutputPath
-              .resolveSibling(
-                  "temp-beam-"
-                  + baseOutputPath.getFileName()
-                  + "-"
-                  + now.toString(DateTimeFormat.forPattern("yyyy-MM-DD_HH-mm-ss")))
-              .toString();
-        } catch (IOException e) {
-          throw new RuntimeException(e);
-        }
+        URI baseOutputUri = URI.create(baseOutputFilename).normalize();
```

Comment:
* This fragment looks like something that should a single call to PathUtils - AFAIU the whole point of PathUtils is to free clients from this error-prone string and URI manipulation.
* I don't understand what's going on here in general:
** does URI.create() work for Windows paths (e.g. "c:\foo bar\blah")?
** What's the difference between baseOutputUri and pathUri?
** What does the self-relativize in "pathUri.resolve("").relativize(pathUri)" accomplish?
** Why do we hand-concatenate paths at line 332 instead of using .resolve()?
* Line 323 says we shouldn't call Instant.now() inside .apply(), but at line 334 we're calling it - I'm surprised there's no test for that, can you add one?